crypto/armcap.c

```diff
@@ -174,11 +179,18 @@ void OPENSSL_cpuid_setup(void)
      */
 #   else
     {
-        unsigned int sha512;
-        size_t len = sizeof(sha512);
+        unsigned int feature;
+        size_t len = sizeof(feature);
+        char uarch[64];
 
-        if (sysctlbyname("hw.optional.armv8_2_sha512", &sha512, &len, NULL, 0) == 0 && sha512 == 1)
+        if (sysctlbyname("hw.optional.armv8_2_sha512", &feature, &len, NULL, 0) == 0 && feature == 1)
             OPENSSL_armcap_P |= ARMV8_SHA512;
+        feature = 0;
+        if (sysctlbyname("hw.optional.armv8_2_sha3", &feature, &len, NULL, 0) == 0 && feature == 1)
+            OPENSSL_armcap_P |= ARMV8_SHA3;
+        len = sizeof(uarch);
+        if (sysctlbyname("machdep.cpu.brand_string", uarch, &len, NULL, 0) == 0)
+            m1_flag = (strcmp(uarch, "Apple M1") == 0);
```

Comment:
comparing the name of the brand looks fragile for a feature detection

Reply:
Hi， the m1_flag, as one local variable in the function OPENSSL_cpuid_setup, is only used  to decide whether we enable OPENSSL_armv8_aes_unroll8_eor3 or not.  and the micro-architecture detection function highly depend on OPENSSL_armv8_aes_unroll8_eor3 other than m1_flag. So, should we change the variable name m1_flag to other?

Reply:
How about your option? @bernd-edlinger

Reply:
@bernd-edlinger Here `machdep.cpu.brand_string` is being used for microarchitecture detection, rather than feature detection, and is actually a conservative test to only bring in the new code when this exact CPU is reported. M1 processors support the crypto extensions needed here - they aren't available via `sysctl`, so most software (e.g. clang, golang) is treating them as non-optional like this code does

Reply:
please define a bit in OPENSSL_armcap_P and for each of these features.
Set the bit if you detect the CPU brand, but allow the user to override this detection, like all other feature bits.

Reply:
We can see the features in OPENSSL_armcap_P  as below:
```
 # define ARMV7_NEON      (1<<0)
# define ARMV7_TICK      (1<<1)
# define ARMV8_AES       (1<<2)
# define ARMV8_SHA1      (1<<3)
# define ARMV8_SHA256    (1<<4)
# define ARMV8_PMULL     (1<<5)
# define ARMV8_SHA512    (1<<6)
```

each identifies one feature, it seems wired if we list the cpu brand bit into them.  
How about your opinion?  thanks for the further comments.

Reply:
Sorry, but I see this as a feature.
Maybe you have to invent a name for whatever_CPU_brand_X_can_do first.

Reply:
A feature is "the CPU has this instruction" - it's either true or false, and so useful to set bits for (and to allow override for).

Microarchitectural detection is "what CPU is this?", because whether a particular unroll count or other set of optimisations etc is better or worse will depend on that. It's not a "can this CPU do this?", but a "does this set of techniques work better on this exact CPU or not?". And as we develop this code further, different CPUs will do better or worse with particular optimisations.

CPU microarchitecture feeds into potentially several different sets of optimisations, and isn't really an on/off-type question, so bits don't seem useful. If it's important to be able to provide an override, we could add support for an environment variable something like `OPENSSL_armcap_cpuid` and then not do *any* microarchitectural detection if it's specified.

If you'd rather bits that can be set on/off, then they would be "do optimisation X", "do optimisation Y", etc, which seems a little fine-grained for most users.